File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_py/BinaryDataProcessor.py
class BinaryInfo:
    def __init__(self, zeroes, ones, bit_length):
        self.zeroes = zeroes
        self.ones = ones
        self.bit_length = bit_length

    def __str__(self):
        return f"{{Zeroes: {self.zeroes:.3f}, Ones: {self.ones:.3f}, Bit length: {self.bit_length}}}"

class BinaryDataProcessor:
    def __init__(self, binary_string):
        self.binary_string = binary_string
        self.clean_non_binary_chars()

    def clean_non_binary_chars(self):
        self.binary_string = ''.join(c for c in self.binary_string if c in '01')

    def calculate_binary_info(self):
        total_length = len(self.binary_string)
        zeroes_count = self.binary_string.count('0')
        ones_count = self.binary_string.count('1')
        zeroes_percentage = zeroes_count / total_length
        ones_percentage = ones_count / total_length
        return BinaryInfo(zeroes_percentage, ones_percentage, total_length)

    def convert_to_ascii(self):
        ascii_string = ""
        n = len(self.binary_string)
        for i in range(0, n, 8):
            chunk = self.binary_string[i:i+8]
            if not chunk:
                break
            decimal = int(chunk, 2)
            ascii_string += chr(decimal)
        return ascii_string

    def convert_to_utf8(self):
        utf8_string = ""
        n = len(self.binary_string)
        for i in range(0, n, 8):
            chunk = self.binary_string[i:i+8]
            if not chunk:
                break
            decimal = int(chunk, 2)
            utf8_string += chr(decimal)
        return utf8_string

    def get_binary_string(self):
        return self.binary_string

if __name__ == "__main__":
    bdp = BinaryDataProcessor("0110100001100101011011000110110001101111")
    print(bdp.get_binary_string())
    print(bdp.calculate_binary_info())
    print(bdp.convert_to_ascii())
    print(bdp.convert_to_utf8())

File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/BinaryDataProcessor.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <cstdio>
#include <stdexcept>

class BinaryInfo {
private:
    double zeroes;
    double ones;
    int bitLength;

public:
    BinaryInfo(double zeroes, double ones, int bitLength) : zeroes(zeroes), ones(ones), bitLength(bitLength) {}

    double getZeroes() const { return zeroes; }
    double getOnes() const { return ones; }
    int getBitLength() const { return bitLength; }

    std::string toString() const {
        char buffer[100];
        snprintf(buffer, sizeof(buffer), "{Zeroes: %.3f, Ones: %.3f, Bit length: %d}", zeroes, ones, bitLength);
        return std::string(buffer);
    }
};

class BinaryDataProcessor {
private:
    std::string binaryString;

    void cleanNonBinaryChars() {
        auto it = std::remove_if(binaryString.begin(), binaryString.end(),
            [](char c) { return c != '0' && c != '1'; });
        binaryString.erase(it, binaryString.end());
    }

public:
    BinaryDataProcessor(std::string binaryString) : binaryString(std::move(binaryString)) {
        cleanNonBinaryChars();
    }

    BinaryInfo calculateBinaryInfo() {
        int totalLength = binaryString.length();
        int zeroesCount = std::count(binaryString.begin(), binaryString.end(), '0');
        int onesCount = totalLength - zeroesCount;

        double zeroesPercentage = static_cast<double>(zeroesCount) / totalLength;
        double onesPercentage = static_cast<double>(onesCount) / totalLength;

        return BinaryInfo(zeroesPercentage, onesPercentage, totalLength);
    }

    std::string convertToAscii() {
        std::string asciiString;
        for (size_t i = 0; i < binaryString.length(); i += 8) {
            if (i + 8 > binaryString.length()) {
                throw std::out_of_range("Index out of bounds");
            }
            std::string byteString = binaryString.substr(i, 8);
            int decimal = std::stoi(byteString, nullptr, 2);
            asciiString.push_back(static_cast<char>(decimal));
        }
        return asciiString;
    }

    std::string convertToUtf8() {
        std::string utf8String;
        for (size_t i = 0; i < binaryString.length(); i += 8) {
            if (i + 8 > binaryString.length()) {
                throw std::out_of_range("Index out of bounds");
            }
            std::string byteString = binaryString.substr(i, 8);
            int decimal = std::stoi(byteString, nullptr, 2);
            utf8String.push_back(static_cast<char>(decimal));
        }
        return utf8String;
    }

    std::string getBinaryString() const {
        return binaryString;
    }
};

int main() {
    BinaryDataProcessor bdp("0110100001100101011011000110110001101111");
    std::cout << bdp.getBinaryString() << std::endl;
    BinaryInfo info = bdp.calculateBinaryInfo();
    std::cout << info.toString() << std::endl;
    std::cout << bdp.convertToAscii() << std::endl;
    std::cout << bdp.convertToUtf8() << std::endl;
    return 0;
}